bin/embark

```diff
@@ -1,15 +1,1023 @@
 #!/usr/bin/env node
 
-try {
-  eval('let __nodeTest = 123;');
-} catch(e) {
-  if (e.name === 'SyntaxError') {
-    console.error("unsupported version of NodeJS. Make sure you are running nodejs 8.11.3 or above");
+/* global __dirname __filename process require */
 
-    process.exit();
+// This script doesn't use JS syntax, packages, or APIs *un*supported by any
+// node version >=4.0.0, so unsupported versions from v4.0.0+ will get the
+// intended error messages. A future version of this script could instead rely
+// on babel to achieve the same goal.
+// See: https://node.green/
+
+// KEY ASSUMPTION: for a DApp to be valid, from embark's perspective, it must
+// have a parsable embark.json file in its top-level directory; if that
+// requirement changes in the future then this script must be revised.
+// Hypothetical example of such a change: embark config info may be included in
+// package.json under `{"embark": {...}}` -or- stored in an embark.json file.
+
+function main() {
+  whenNoShim();
+  var invoked = thisEmbark();
+  var embarkJson = findEmbarkJson();
+  var dappPath = embarkJson.dirname;
+  process.chdir(dappPath);
+  process.env.DAPP_PATH = dappPath;
+  process.env.PWD = dappPath;
+
+  /* attempt to find a "local" embark in or above but not below dappPath
+
+     let `dappPath/(([../])*)bin/embark` be a "containing" embark
+
+     let `dappPath/(([../])*)node_modules/embark/bin/embark` be an "installed"
+     embark
+
+     if containing and installed embarks are both found, and if containing
+     embark is higher in the dir structure than installed embark, then
+     containing embark will be selected
+
+     according to the rule above: if an installed embark is found within a
+     containing embark's own node_modules (that would be odd), installed embark
+     will be selected
+
+     invoked embark may find itself as local embark, but that is detected by
+     comparing `binrealpath` props to avoid double-checking and infinite loops
+
+     if no local embark is found then cmd execution will use invoked embark */
+
+  var containing = findBinContaining(dappPath, invoked);
+  var installed = findBinInstalled(dappPath, invoked);
+  var local = selectLocal(containing, installed);
+  var pkgJson = findPkgJson(dappPath, embarkJson, local);
+  process.env.PKG_PATH = pkgJson.dirname;
+  var embark = select(invoked, local);
+  process.env.EMBARK_PATH = embark.pkgDir;
+  embark.exec();
+}
+
+// -----------------------------------------------------------------------------
+
+var checkDeps = require('check-dependencies');
+var npmlog = require('npmlog');
+var findUp = require('find-up');
+var fs = require('fs');
+var path = require('path');
+var parseJsonWithErrors = require('json-parse-better-errors');
+var pkgUp = require('pkg-up');
+var semver = require('semver');
+var subdir = require('subdir');
+
+// -- embark bins --------------------------------------------------------------
```

Comment:
Regarding comments in general, I understand why you did it (complexity).
Should we not try to use a parsable format like jsDoc?

That would make the comments more valuable I believe

Reply:
It's a good idea. I'm not familiar with the jsDoc format (I know what it is, have never learned it). Do you think that for this PR it's worth the time for me to learn the basics and make some revisions? Or is it an improvement that could/should come in a later PR?

Reply:
Up to you, I don't want to block the PR, something we could add later indeed.
@team what do you think?